README.md
[![Build](https://github.com/jthomperoo/horizontal-pod-autoscaler/workflows/main/badge.svg)](https://github.com/jthomperoo/horizontal-pod-autoscaler/actions)
[![go.dev](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white&style=flat)](https://pkg.go.dev/github.com/jthomperoo/horizontal-pod-autoscaler)
[![Go Report Card](https://goreportcard.com/badge/github.com/jthomperoo/horizontal-pod-autoscaler)](https://goreportcard.com/report/github.com/jthomperoo/horizontal-pod-autoscaler)
[![Documentation Status](https://readthedocs.org/projects/predictive-horizontal-pod-autoscaler/badge/?version=latest)](https://predictive-horizontal-pod-autoscaler.readthedocs.io/en/latest)
[![License](https://img.shields.io/:license-apache-blue.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)

# Custom Pod Autoscaler - Horizontal Pod Autoscaler (CPA-HPA)

This is the Horizontal Pod Autoscaler (HPA), modified to work as a [Custom Pod Autoscaler
(CPA)](https://github.com/jthomperoo/custom-pod-autoscaler). This project is designed to be a starting point to allow
developers to quickly take a working Horizontal Pod Autoscaler and modify it as they need.

## What is a Custom Pod Autoscaler?

A Custom Pod Autoscaler is a way to write custom logic into Kubernetes scalers.
For more detailed overview, see the [Custom Pod Autoscaler
Framework](https://custom-pod-autoscaler.readthedocs.io/en/latest/).

## How do I use this project?

If you want to deploy this CPA-HPA onto your cluster, you first need to install the [Custom Pod Autoscaler
Operator](https://github.com/jthomperoo/custom-pod-autoscaler-operator), follow the [installation guide for
instructions for installing the
operator](https://github.com/jthomperoo/custom-pod-autoscaler-operator/blob/master/INSTALL.md).

## Overview of codebase

This project should be functionally similar/identical to the Kubernetes Horizontal Pod Autoscaler, with differences
being that this is run as a CPA rather than a Kubernetes controller. The code is largely copied from the Kubernetes
HPA and modified to work with the Custom Pod Autoscaler. Some restructuring of the architecture was required to fit
with how a CPA operates, with the HPA logic now split into two distinct parts; metric gathering and evaluation.

### Metric Gathering

The metric gathering stage takes in a resource to gather metrics for, and metric spec definitions for which metrics
to gather. Using this information the metrics are gathered and calculated, from metrics APIs. These metrics are then
serialised into JSON and output back to the CPA through stdout.

### Evaluation

The evaluation stage takes in the metrics gathered by the metric gathering stage and makes a decision - how many
replicas should the resource have. Once this decision has been made using data available the decision is serialised
into JSON and output back to the CPA through stdout to do the actual scaling.

### Configuration

There is some overlap in functionality between the the CPA and HPA, in this project precedence has been given to the
CPA functionality. For example, Kubernetes HPAs have the ability to set minimum and maximum replica counts, this
project removes those and instead relies on the CPA minimum and maximum replica counts. A Kubernetes HPA also contains
a ScaleTargetRef for deciding which resource to target when scaling; for this project only the ScaleTargetRef of the
CPA is used.

Deciding which metrics to use is done by using `MetricSpecs`, which are a key part of HPAs, and look like this:

```yaml
- type: Resource
  resource:
    name: cpu
    target:
        type: Utilization
        averageUtilization: 50
```

To send these specs to the CPA-HPA, add a config option called `metrics` to the CPA, with a multiline string
containing the metric list. For example:

```yaml
config:
  - name: metrics
    value: |
      - type: Resource
        resource:
          name: cpu
          target:
            type: Utilization
            averageUtilization: 50
```

## Example

There is an example of how to use the Custom Pod Autoscaler - Horizontal Pod Autoscaler in [`/example`](./example).
The example has a simple deployment, taken from the [Kubernetes Horizontal Pod Autoscaler
walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/). The example also
contains the YAML definition of the CPA-HPA.

## More information

See the [wiki for more information, such as guides and
references](https://horizontal-pod-autoscaler.readthedocs.io/en/latest/).

## Developing this project
### Environment
Developing this project requires these dependencies:

* [Go](https://golang.org/doc/install) >= 1.17
* [staticcheck](https://staticcheck.io/docs/getting-started/) == `v0.3.0 (2022.1)`
* [Docker](https://docs.docker.com/install/)

To view the docs, you need Python 3 installed:

* [Python](https://www.python.org/downloads/) >= `3.8`

To view docs locally you need some Python dependencies, run:

```bash
pip install -r docs/requirements.txt
```

### Commands

* `make` - builds the CPA-HPA binary.
* `make docker` - builds the CPA-HPA image.
* `make lint` - lints the code.
* `make format` - formats the code.
* `make test` - runs the project tests.
* `make doc` - hosts the documentation locally, at `127.0.0.1:8000`.
* `make coverage` - shows the test coverage report.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.9.0] - 2022-05-14
### Changed
- Bump `jthomperoo/custom-pod-autoscaler` to `v2.6.0`.
- Extracted all HPA calculations to separate project <https://github.com/jthomperoo/k8shorizmetrics>.
- Updated Go version to `v1.17`.
### Removed
- Removed all public modules from this project, if depending on this instead use <https://github.com/jthomperoo/k8shorizmetrics>.

## [v0.8.0] - 2021-12-27
### Added
- Support for `argoproj.io/v1alpha1` `Rollout` resource.
### Changed
- Bump `jthomperoo/custom-pod-autoscaler` to `v2.3.0`
- Bump `k8s.io/kubernetes` to `v1.21.8`.
### Fixed
- **BREAKING CHANGE**, the Pod metric evaluation now correctly looks at the `averageValue` metric target field rather
than the incorrect `value` target field.

## [v0.7.0] - 2021-08-15
### Changed
- Update Custom Pod Autoscaler version to `v2.2.0`.
- Update Go version to `v1.16`.

## [v0.6.0] - 2020-08-31
### Changed
- Update Custom Pod Autoscaler version to `v1.0.0`.

## [v0.5.0] - 2020-03-10
### Changed
- Update Custom Pod Autoscaler version to `v0.11.0`.
- Changed `cpuInitializationPeriod`, time now set in seconds rather than minutes.
- Set default `interval` to be `15000` (15 seconds) to match K8s HPA.
- Set default `downscaleStabilization` to be `300` (5 minutes) to match K8s HPA.

## [v0.4.0] - 2020-01-25
### Changed
- Update Custom Pod Autoscaler version to v0.10.0.

## [v0.3.0] - 2019-12-17
### Changed
- Update Custom Pod Autoscaler version to 0.8.0.

## [v0.2.0] - 2019-12-10
### Added
- New configuration options can be set in the YAML config.
    - Can now configure `tolerance` as a configuration option - works the same as the `--horizontal-pod-autoscaler-tolerance` flag, [see here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). Default value of 0.1.
    - Can now configure `cpuInitializationPeriod` as a configuration option - works the same as the `--horizontal-pod-autoscaler-cpu-initialization-period` flag, [see here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). Time set in minutes, default 5 minutes.
    - Can now configure `initialReadinessDelay` as a configuration option - works the same as the `--horizontal-pod-autoscaler-initial-readiness-delay` flag, [see here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). Time set in seconds, default 30 seconds.
### Fixed
- Issues with evaluation decision making looking in the wrong specs for target values.

## [v0.1.0] - 2019-12-08
### Added
- Restructured the Horizontal Pod Autoscaler to work within a Custom Pod Autoscaler.

[Unreleased]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.9.0...HEAD
[v0.9.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.8.0...v0.9.0
[v0.8.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.7.0...v0.8.0
[v0.7.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.6.0...v0.7.0
[v0.6.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.5.0...v0.6.0
[v0.5.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.4.0...v0.5.0
[v0.4.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.3.0...v0.4.0
[v0.3.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/jthomperoo/horizontal-pod-autoscaler/releases/tag/v0.1.0


main.go
/*
Copyright 2021 The Custom Pod Autoscaler Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Horizontal Pod Autoscaler provides executable Horizontal Pod Autoscaler logic, which
// can be built into a Custom Pod Autoscaler.
// The Horizontal Pod Autoscaler has two modes, metric gathering and evaluation.
// Metric mode gathers metrics, taking in a resource to get the metrics for and outputting
// these metrics as serialised JSON.
// Evaluation mode makes decisions on how many replicas a resource should have, taking in
// metrics and outputting evaluation decisions as seralised JSON.
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"strings"
	"time"

	autoscalingv1 "k8s.io/api/autoscaling/v1"

	cpaevaluate "github.com/jthomperoo/custom-pod-autoscaler/v2/evaluate"
	cpametric "github.com/jthomperoo/custom-pod-autoscaler/v2/metric"
	"github.com/jthomperoo/k8shorizmetrics"
	"github.com/jthomperoo/k8shorizmetrics/metrics"
	"github.com/jthomperoo/k8shorizmetrics/metricsclient"
	"github.com/jthomperoo/k8shorizmetrics/podsclient"
	autoscalingv2 "k8s.io/api/autoscaling/v2beta2"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/labels"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/apimachinery/pkg/util/yaml"
	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/restmapper"
	k8sscale "k8s.io/client-go/scale"
)

const (
	defaultTolerance = float64(0.1)
	// 5 minute CPU initialization period
	defaultCPUInitializationPeriod = 300
	// 30 second initial readiness delay
	defaultInitialReadinessDelay = 30
)

// EvaluateSpec represents the information fed to the evaluator
type EvaluateSpec struct {
	Metrics              []*cpametric.ResourceMetric `json:"metrics"`
	UnstructuredResource unstructured.Unstructured   `json:"resource"`
	RunType              string                      `json:"runType"`
}

// MetricSpec represents the information fed to the metric gatherer
type MetricSpec struct {
	UnstructuredResource unstructured.Unstructured `json:"resource"`
	RunType              string                    `json:"runType"`
}

func main() {
	stdin, err := ioutil.ReadAll(os.Stdin)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	modePtr := flag.String("mode", "no_mode", "command mode, either metric or evaluate")
	flag.Parse()

	switch *modePtr {
	case "metric":
		gather(bytes.NewReader(stdin))
	case "evaluate":
		evaluate(bytes.NewReader(stdin))
	default:
		log.Fatalf("Unknown command mode: %s", *modePtr)
		os.Exit(1)
	}
}

func gather(stdin io.Reader) {
	var spec MetricSpec
	err := yaml.NewYAMLOrJSONDecoder(stdin, 10).Decode(&spec)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	metricSpecsValue, exists := os.LookupEnv("metrics")
	if !exists {
		log.Fatal("Metric specs not supplied")
		os.Exit(1)
	}

	// Read in metric specs to evaluate
	var metricSpecs []autoscalingv2.MetricSpec
	err = yaml.NewYAMLOrJSONDecoder(strings.NewReader(metricSpecsValue), 10).Decode(&metricSpecs)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	if len(metricSpecs) == 0 {
		log.Fatal("Metric specs not supplied")
		os.Exit(1)
	}

	// Get initial readiness delay, can be set as a configuration variable
	initialReadinessDelaySecs, err := parseInt64EnvVar("initialReadinessDelay", defaultInitialReadinessDelay)
	if err != nil {
		log.Fatalf("Invalid initial readiness delay provided: %e\n", err)
		os.Exit(1)
	}

	// Get CPU initialization period, can be set as a configuration variable
	cpuInitializationPeriodSecs, err := parseInt64EnvVar("cpuInitializationPeriod", defaultCPUInitializationPeriod)
	if err != nil {
		log.Fatalf("Invalid CPU initialization period provided: %e\n", err)
		os.Exit(1)
	}

	clusterConfig, clientset, err := getKubernetesClients()
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	scale, err := getScaleSubResource(clientset, &spec.UnstructuredResource)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	metricsclient := metricsclient.NewClient(clusterConfig, clientset.Discovery())
	podsclient := &podsclient.OnDemandPodLister{
		Clientset: clientset,
	}
	cpuInitializationPeriod := time.Duration(cpuInitializationPeriodSecs) * time.Second
	initialReadinessDelay := time.Duration(initialReadinessDelaySecs) * time.Second

	// Create metric gatherer, with required clients and configuration
	gatherer := k8shorizmetrics.NewGatherer(metricsclient, podsclient, cpuInitializationPeriod, initialReadinessDelay)

	selector, err := labels.Parse(scale.Status.Selector)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// Get metrics for deployment
	metrics, err := gatherer.Gather(metricSpecs, scale.GetNamespace(), selector)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// Marshal metrics into JSON
	jsonMetrics, err := json.Marshal(metrics)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// Write serialised metrics to stdout
	fmt.Print(string(jsonMetrics))
}

func evaluate(stdin io.Reader) {
	var spec EvaluateSpec
	err := yaml.NewYAMLOrJSONDecoder(stdin, 10).Decode(&spec)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// Get tolerance, can be set as a configuration variable
	var tolerance float64
	toleranceValue, exists := os.LookupEnv("tolerance")
	if !exists {
		// use default
		tolerance = defaultTolerance
	} else {
		// try to parse provided value
		tolerance, err = strconv.ParseFloat(toleranceValue, 64)
		if err != nil {
			log.Fatalf("Invalid tolerance provided: %e\n", err)
			os.Exit(1)
		}
	}

	_, clientset, err := getKubernetesClients()
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	scale, err := getScaleSubResource(clientset, &spec.UnstructuredResource)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	var combinedMetrics []*metrics.Metric
	err = yaml.NewYAMLOrJSONDecoder(strings.NewReader(spec.Metrics[0].Value), 10).Decode(&combinedMetrics)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	evaluator := k8shorizmetrics.NewEvaluator(tolerance)
	targetReplicas, err := evaluator.Evaluate(combinedMetrics, scale.Spec.Replicas)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// Marshal evaluation into JSON
	jsonEvaluation, err := json.Marshal(cpaevaluate.Evaluation{
		TargetReplicas: targetReplicas,
	})
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	// Write serialised evaluation to stdout
	fmt.Print(string(jsonEvaluation))
}

func getScaleSubResource(clientset *kubernetes.Clientset, resource *unstructured.Unstructured) (*autoscalingv1.Scale, error) {
	groupResources, err := restmapper.GetAPIGroupResources(clientset.Discovery())
	if err != nil {
		return nil, err
	}

	scaleClient := k8sscale.New(
		clientset.RESTClient(),
		restmapper.NewDiscoveryRESTMapper(groupResources),
		dynamic.LegacyAPIPathResolverFunc,
		k8sscale.NewDiscoveryScaleKindResolver(
			clientset.Discovery(),
		),
	)

	resourceGV, err := schema.ParseGroupVersion(resource.GetAPIVersion())
	if err != nil {
		return nil, err
	}

	targetGR := schema.GroupResource{
		Group:    resourceGV.Group,
		Resource: resource.GetKind(),
	}

	scale, err := scaleClient.Scales(resource.GetNamespace()).Get(context.Background(), targetGR, resource.GetName(), metav1.GetOptions{})
	if err != nil {
		return nil, err
	}

	return scale, nil
}

func getKubernetesClients() (*rest.Config, *kubernetes.Clientset, error) {
	clusterConfig, err := rest.InClusterConfig()
	if err != nil {
		return nil, nil, err
	}

	clientset, err := kubernetes.NewForConfig(clusterConfig)
	if err != nil {
		return nil, nil, err

	}

	return clusterConfig, clientset, nil
}

func parseInt64EnvVar(name string, defaultValue int64) (int64, error) {
	var envVar int64
	var err error
	envVarVal, exists := os.LookupEnv(name)
	if !exists {
		// use default
		envVar = defaultValue
	} else {
		// try to parse provided value
		envVar, err = strconv.ParseInt(envVarVal, 10, 64)
		if err != nil {
			return 0, err
		}
	}
	return envVar, nil
}
